integration/sandbox_run_rollback_test.go

```diff
@@ -229,6 +235,97 @@ func TestRunPodSandboxWithShimStartAndTeardownCNIFailure(t *testing.T) {
 	t.Run("JustCleanup", testCase(false))
 }
 
+// TestRunPodSandboxWithShimStartAndTeardownCNISlow should keep the sandbox
+// record if failed to rollback CNI API.
+func TestRunPodSandboxWithShimStartAndTeardownCNISlow(t *testing.T) {
+	if runtime.GOOS != "linux" {
+		t.Skip()
+	}
+	if os.Getenv("ENABLE_CRI_SANDBOXES") != "" {
+		t.Skip()
+	}
+
+	t.Log("Init PodSandboxConfig with specific key")
+	labels := map[string]string{
+		t.Name(): "true",
+	}
+	sbConfig := PodSandboxConfig(t.Name(), "failpoint", WithPodLabels(labels))
+
+	t.Log("Inject CNI failpoint")
+	conf := &failpointConf{
+		// Delay 10s
+		Add: "1*delay(10000)",
+		// This is to make sure the sandbox container won't be deleted even if the defer functions are invoked in sandbox_run.go
+		// Without this, the test is flaky because ListPodSandbox may return empty list.
+		Del: "1*error(please retry)",
+	}
+	injectCNIFailpoint(t, sbConfig, conf)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+		t.Log("Create a sandbox")
+		_, err := runtimeService.RunPodSandbox(sbConfig, failpointRuntimeHandler)
+		require.Error(t, err)
+		require.ErrorContains(t, err, "error reading from server: EOF")
+	}()
+
+	time.Sleep(5 * time.Second)
```

Comment:
The sleep will make the case flaky. 
For instance, if it takes long time to download the sandbox image, the 5s is not enough.

I suggest to scan the procfs to check the cni.add is running.
Here is my patch https://github.com/fuweid/containerd/commit/adc7b4404b2fe20ce91e29263b752f9568bf17d7

```diff
commit adc7b4404b2fe20ce91e29263b752f9568bf17d7
Author: Wei Fu <fuweid89@gmail.com>
Date:   Fri Oct 7 15:06:54 2022 +0800

    integration: scan procfs to check the CNI is running

    Signed-off-by: Wei Fu <fuweid89@gmail.com>

diff --git a/integration/main_test.go b/integration/main_test.go
index 96fe33769..d56179af3 100644
--- a/integration/main_test.go
+++ b/integration/main_test.go
@@ -17,11 +17,13 @@
 package integration

 import (
+       "bytes"
        "context"
        "encoding/json"
        "errors"
        "flag"
        "fmt"
+       "io"
        "os"
        "path/filepath"
        goruntime "runtime"
@@ -445,6 +447,80 @@ func PidOf(name string) (int, error) {
        return strconv.Atoi(output)
 }

+// PidsOf returns pid(s) of a process by name
+func PidsOf(name string) ([]int, error) {
+       if len(name) == 0 {
+               return []int{}, fmt.Errorf("name is required")
+       }
+
+       procDirFD, err := os.Open("/proc")
+       if err != nil {
+               return nil, fmt.Errorf("failed to open /proc: %w", err)
+       }
+       defer procDirFD.Close()
+
+       res := []int{}
+       for {
+               fileInfos, err := procDirFD.Readdir(100)
+               if err != nil {
+                       if err == io.EOF {
+                               break
+                       }
+                       return nil, fmt.Errorf("failed to readdir: %w", err)
+               }
+
+               for _, fileInfo := range fileInfos {
+                       if !fileInfo.IsDir() {
+                               continue
+                       }
+
+                       pid, err := strconv.Atoi(fileInfo.Name())
+                       if err != nil {
+                               continue
+                       }
+
+                       exePath, err := os.Readlink(filepath.Join("/proc", fileInfo.Name(), "exe"))
+                       if err != nil {
+                               continue
+                       }
+
+                       if strings.HasSuffix(exePath, name) {
+                               res = append(res, pid)
+                       }
+               }
+       }
+       return res, nil
+}
+
+// PidEnvs returns the environ of pid in key-value pairs.
+func PidEnvs(pid int) (map[string]string, error) {
+       envPath := filepath.Join("/proc", strconv.Itoa(pid), "environ")
+
+       b, err := os.ReadFile(envPath)
+       if err != nil {
+               return nil, fmt.Errorf("failed to read %s: %w", envPath, err)
+       }
+
+       values := bytes.Split(b, []byte{0})
+       if len(values) == 0 {
+               return nil, nil
+       }
+
+       res := make(map[string]string)
+       for _, value := range values {
+               value := strings.TrimSpace(string(value))
+               if len(value) == 0 {
+                       continue
+               }
+
+               parts := strings.SplitN(value, "=", 2)
+               if len(parts) == 2 {
+                       res[parts[0]] = parts[1]
+               }
+       }
+       return res, nil
+}
+
 // RawRuntimeClient returns a raw grpc runtime service client.
 func RawRuntimeClient() (runtime.RuntimeServiceClient, error) {
        addr, dialer, err := dialer.GetAddressAndDialer(*criEndpoint)
diff --git a/integration/sandbox_run_rollback_test.go b/integration/sandbox_run_rollback_test.go
index 5cb13c3ba..cba2a8b86 100644
--- a/integration/sandbox_run_rollback_test.go
+++ b/integration/sandbox_run_rollback_test.go
@@ -42,6 +42,7 @@ import (

 const (
        failpointRuntimeHandler = "runc-fp"
+       failpointCNIBinary      = "cni-bridge-fp"

        failpointShimPrefixKey = "io.containerd.runtime.v2.shim.failpoint."

@@ -246,18 +247,17 @@ func TestRunPodSandboxWithShimStartAndTeardownCNISlow(t *testing.T) {
        }

        t.Log("Init PodSandboxConfig with specific key")
+       sbName := t.Name()
+
        labels := map[string]string{
-               t.Name(): "true",
+               sbName: "true",
        }
-       sbConfig := PodSandboxConfig(t.Name(), "failpoint", WithPodLabels(labels))
+       sbConfig := PodSandboxConfig(sbName, "failpoint", WithPodLabels(labels))

        t.Log("Inject CNI failpoint")
        conf := &failpointConf{
-               // Delay 10s
-               Add: "1*delay(10000)",
-               // This is to make sure the sandbox container won't be deleted even if the defer functions are invoked in sandbox_run.go
-               // Without this, the test is flaky because ListPodSandbox may return empty list.
-               Del: "1*error(please retry)",
+               // Delay 1 day
+               Add: "1*delay(86400000)",
        }
        injectCNIFailpoint(t, sbConfig, conf)

@@ -273,8 +273,40 @@ func TestRunPodSandboxWithShimStartAndTeardownCNISlow(t *testing.T) {
                require.ErrorContains(t, err, "error reading from server: EOF")
        }()

-       time.Sleep(5 * time.Second)
-       t.Log("Restart containerd after 5s")
+       assert.NoError(t, Eventually(func() (bool, error) {
+               pids, err := PidsOf(failpointCNIBinary)
+               if err != nil || len(pids) == 0 {
+                       return false, err
+               }
+
+               for _, pid := range pids {
+                       envs, err := PidEnvs(pid)
+                       if err != nil {
+                               t.Logf("failed to read environ of pid %v: %v: skip it", pid, err)
+                               continue
+                       }
+
+                       args, ok := envs["CNI_ARGS"]
+                       if !ok {
+                               t.Logf("expected CNI_ARGS env but got nothing, skip pid=%v", pid)
+                               continue
+                       }
+
+                       for _, arg := range strings.Split(args, ";") {
+                               kv := strings.SplitN(arg, "=", 2)
+                               if len(kv) != 2 {
+                                       continue
+                               }
+
+                               if kv[0] == "K8S_POD_NAME" && kv[1] == sbName {
+                                       return true, nil
+                               }
+                       }
+               }
+               return false, nil
+       }, time.Second, 30*time.Second), "check that failpoint CNI.Add is running")
+
+       t.Log("Restart containerd")
        RestartContainerd(t)

        wg.Wait()
```

Reply:
Thank you for the suggestion. That sounds better than sleeping a random time. I will take it.